designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,372 @@
+  disruption:
+    consolidationPolicy: WhenUnderutilized 
+    expireAfter: Never 
+    budgets:
+    - nodes: 0
+      reasons: [Drifted] 
+    - nodes: "100%" // All Disruption Other than drift(Expiration and Consolidation) are allowed
+```
+
+#### Fully Blocking Drift and Expiration while allowing emptiness 
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenEmpty
+    expireAfter: Never 
+    budgets:
+    - nodes: 0
+      reasons: [Drifted] 
+    - nodes: "100%" // All Disruption Other than drift(Expiration and Empty Nodes) are allowed
+```
+
+#### Limiting Drift & Expiration while allowing aggressive but not unbounded consolidation
+Drift and Expiration may not be desired reasons to remove many nodes in one interval, but Underutilization and Emptiness may be desired. We may want to still consolidate at a healthy threshold but not allow 
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized 
+    expireAfter: 1h 
+    budgets:
+    - nodes: 1
+      reasons: [Drifted, Expiration] # Only allow Drift and Expiration to occur one at a time 
+    - nodes: "33%" # All Consolidation actions taken either through replace or delete should be allowed to disrupt at least a third of the cluster at once 
+```
+
+#### Limiting Cost Saving disruption to occur during weekends 
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized 
+    expireAfter: 720h 
+    budgets:
+    - nodes: "33%"
+      reasons: [Drifted, Expiration] # Allow Large Drift and Expiration at All Times as these are required for cluster health 
+    - nodes: "0" # All Consolidation actions taken either through replace or delete should be allowed to disrupt at least a third of the cluster at once 
+      schedule: "* * * * mon-fri" 
+      duration: 24h 
+      reasons: ["underutilized"] 
+    - nodes: "100%" 
+      schedule: "* * * * sat"
+      duration: 48h
+      reasons: ["underutilized"]
+```
+
+#### Which Equation Best Covers all the scenarios? 
+### Q: How should we handle an unspecified reason when others are specified? 
+```yaml
+budgets: 
+  - nodes: 10
+    reasons: [Drifted, Underutilized]
+    schedule: "* * * * *"
+  - nodes: 5 
+    reasons: [Empty] 
+    schedule: "* * * * *" 
+```
+In the case of a budget like above, default is undefined. Should karpenter assume the user doesn't want to disrupt any other reasons? Or should we assume that if a default is unspecified, they want us to disrupt anyway?  
+The intuitive options if there is no active default budget is to allow disruption of either 0 or total number of nodes(meaning unbounded disruption).
```

Comment:
I'd expect us to use the current karpenter default, in a case like this that doesn't have a default specified. 

Karpenter's current handling of this is `10%`:
https://karpenter.sh/v0.35/concepts/disruption/#disruption-budgets
![image](https://github.com/kubernetes-sigs/karpenter/assets/33269602/021e54cf-d159-4799-8087-d7fa72d8ec1d)

I think unbound default is both unexpected, and could result in concerning scenarios for unaware users. `10-30`% seems most reasonable, so sticking with the current 10% default seems good to me.